crates/proto/src/domain/account.rs

```diff
@@ -346,37 +351,55 @@ impl From<AccountStorageHeader> for proto::account::AccountStorageHeader {
     }
 }
 
+/// Account vault details
+///
+/// When an account contains a large number of assets (>
+/// [`AccountVaultDetails::MAX_RETURN_ENTRIES`]), including all assets in a single RPC response
+/// creates performance issues. In such cases, the `LimitExceeded` variant indicates to the client
+/// to use the `SyncAccountVault` endpoint instead.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct AccountVaultDetails {
-    pub too_many_assets: bool,
-    pub assets: Vec<Asset>,
+pub enum AccountVaultDetails {
+    /// The vault has too many assets to return inline.
+    /// Clients must use `SyncAccountVault` endpoint instead.
+    LimitExceeded,
+
+    /// The assets in the vault (up to `MAX_RETURN_ENTRIES`).
+    Assets(Vec<Asset>),
 }
+
 impl AccountVaultDetails {
-    const MAX_RETURN_ENTRIES: usize = 1000;
+    /// Maximum number of vault entries that can be returned in a single response.
+    /// Accounts with more assets will have `LimitExceeded` variant.
+    pub const MAX_RETURN_ENTRIES: usize = 1000;
 
     pub fn new(vault: &AssetVault) -> Self {
         if vault.assets().nth(Self::MAX_RETURN_ENTRIES).is_some() {
-            Self::too_many()
+            Self::LimitExceeded
         } else {
-            Self {
-                too_many_assets: false,
-                assets: Vec::from_iter(vault.assets()),
-            }
+            Self::Assets(Vec::from_iter(vault.assets()))
         }
     }
 
     pub fn empty() -> Self {
-        Self {
-            too_many_assets: false,
-            assets: Vec::new(),
+        Self::Assets(Vec::new())
+    }
+
+    /// Creates `AccountVaultDetails` from vault entries (key-value pairs).
+    ///
+    /// This is useful when entries have been fetched directly from the database
+    /// rather than extracted from an `AssetVault`.
+    ///
+    /// The entries are `(vault_key, asset)` pairs where `asset` is a Word representation.
+    pub fn from_entries(entries: Vec<(Word, Word)>) -> Result<Self, miden_objects::AssetError> {
+        if entries.len() > Self::MAX_RETURN_ENTRIES {
+            return Ok(Self::LimitExceeded);
         }
```

Comment:
This is fine for now, but in the future we should avoid reading all the assets from the database. The reason is that we could have vaults with many millions of assets, and it would be pretty costly to read them in, and then discard. To address this, we should probably keep track of the number of assets in the `accounts` table.

Let's create an issue for this.